nab/detectors/htmcore/htmcore_detector.py

```diff
@@ -0,0 +1,268 @@
+      cellsPerColumn            = tmParams["cellsPerColumn"],
+      activationThreshold       = tmParams["activationThreshold"],
+      initialPermanence         = tmParams["initialPerm"],
+      connectedPermanence       = spParams["synPermConnected"],
+      minThreshold              = tmParams["minThreshold"],
+      maxNewSynapseCount        = tmParams["newSynapseCount"],
+      permanenceIncrement       = tmParams["permanenceInc"],
+      permanenceDecrement       = tmParams["permanenceDec"],
+      predictedSegmentDecrement = 0.0,
+      maxSegmentsPerCell        = tmParams["maxSegmentsPerCell"],
+      maxSynapsesPerSegment     = tmParams["maxSynapsesPerSegment"]
+    )
+    self.tm_info = Metrics( [self.tm.numberOfCells()], 999999999 )
+
+    # setup likelihood, these settings are used in NAB
+    if self.useLikelihood:
+      anParams = parameters["anomaly"]["likelihood"]
+      learningPeriod     = int(math.floor(self.probationaryPeriod / 2.0))
+      self.anomalyLikelihood = AnomalyLikelihood(
+                                 learningPeriod= learningPeriod,
+                                 estimationSamples= self.probationaryPeriod - learningPeriod,
+                                 reestimationPeriod= anParams["reestimationPeriod"])
+    # Predictor
+    self.predictor = Predictor( steps=[1, 5], alpha=parameters["predictor"]['sdrc_alpha'] )
+    predictor_resolution = 1
+
+
+  def modelRun(self, ts, val):
+      """
+         Run a single pass through HTM model
+
+         @params ts - Timestamp
+         @params val - float input value
+
+         @return rawAnomalyScore computed for the `val` in this step
+      """
+      ## run data through our model pipeline: enc -> SP -> TM -> Anomaly
+      self.inputs_.append( val )
+      
+      # 1. Encoding
+      # Call the encoders to create bit representations for each value.  These are SDR objects.
+      dateBits        = self.encTimestamp.encode(ts)
+      valueBits       = self.encValue.encode(float(val))
+      # Concatenate all these encodings into one large encoding for Spatial Pooling.
+      encoding = SDR( self.encTimestamp.size + self.encValue.size ).concatenate([valueBits, dateBits])
+      self.enc_info.addData( encoding )
+
+      # 2. Spatial Pooler
+      # Create an SDR to represent active columns, This will be populated by the
+      # compute method below. It must have the same dimensions as the Spatial Pooler.
+      activeColumns = SDR( self.sp.getColumnDimensions() )
+      # Execute Spatial Pooling algorithm over input space.
+      self.sp.compute(encoding, True, activeColumns)
+      self.sp_info.addData( activeColumns )
+
+      # 3. Temporal Memory
+      # Execute Temporal Memory algorithm over active mini-columns.
+      self.tm.compute(activeColumns, learn=True)
+      self.tm_info.addData( self.tm.getActiveCells().flatten() ) #FIXME for anomaly, should we use active cells, or winner cells? And also convert to columns!
```

Comment:
@ctrl-z-9000-times this is taken from hotgym.py. Is this correct, or should: 
- [ ] use `tm.getWinnerCells()` ?
- [ ] convert cells to columns? (I think this with cells is a bug)

Reply:
Non-issue, my mistake. `TM.anomaly` takes care of this.